File: Capitulo_01/01_Getting_Started:_The_Easy_Problems/03_Medium/119_:::_Greedy_Gift_Givers/Main.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;

public class Main
{
	public static void main(String[] args) throws IOException
	{
		BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
		//-----------------------------------------------------------
		Map<String, Integer> m=new LinkedHashMap<String, Integer>();
		int numGrupo, plata, aCuantos, seQueda, seEntrega;
		boolean primero=true;
		
		for(String line, aux[]; (line=br.readLine())!=null;)
		{
			if(!primero) System.out.println();
			else primero=false;
			
			m.clear();
			numGrupo=Integer.parseInt(line);
			aux=br.readLine().split(" ");
			for(String nombre: aux) m.put(nombre, 0);
			
			while(numGrupo--!=0)
			{
				aux=br.readLine().split(" ");
				Integer acum=m.get(aux[0]);
				plata=Integer.parseInt(aux[1]);
				aCuantos=Integer.parseInt(aux[2]);
				
				seEntrega=(aCuantos==0)?0:plata/aCuantos;
				seQueda=plata-seEntrega*aCuantos;
				acum+=seQueda;
				acum-=plata;
				m.put(aux[0], acum);
				
				for(int i=0; i<aCuantos; i++)
				{
					Integer lol=m.get(aux[i+3]);
					lol+=seEntrega;
					m.put(aux[i+3], lol);
				} 
			}
			
			for(Entry<String, Integer> entry : m.entrySet())
				System.out.println(entry.getKey()+" "+entry.getValue());
		}
		//-----------------------------------------------------------
		br.close();
	}
}

File: Capitulo_01/01_Getting_Started:_The_Easy_Problems/03_Medium/119_:::_Greedy_Gift_Givers/Main.c
#include "stdio.h"
#include "string.h"

char mi [10+1][12+1];
int mv[10+1];
int numGrupo;

void putEnMapa(char* aQuien, int acum)
{
	int i;	
	for(i=0; i<numGrupo; i++)
		if(strcmp(aQuien, mi[i])==0)
			mv[i]=acum;
}

int getDelMapa(char* aQuien)
{
	int i;	
	for(i=0; i<numGrupo; i++)
		if(strcmp(aQuien, mi[i])==0)
			return mv[i];
}

int main()
{

	int plata, aCuantos, seQueda, seEntrega, i, lol, acum;
	int primero=1;
	char aQuien [12+1];
	
	while(scanf("%d", &numGrupo) != EOF)
	{
		if(!primero) puts("");
		else primero=0;
		
		for(i=0; i<numGrupo; i++)
		{
			scanf("%s", mi[i]);
			mv[i]=0;	
		}
		
		for(i=0; i<numGrupo; i++)
		{
			scanf("%s %d %d", aQuien, &plata, &aCuantos);
			
			acum=getDelMapa(aQuien);
			seEntrega=(aCuantos==0)?0:plata/aCuantos;
			seQueda=plata-seEntrega*aCuantos;
			acum+=seQueda;
			acum-=plata;
			putEnMapa(aQuien, acum);
			
			while(aCuantos--)
			{
				scanf("%s", aQuien);
				lol=getDelMapa(aQuien);
				lol+=seEntrega;
				putEnMapa(aQuien, lol);
			} 
		}
		
		for(i=0; i<numGrupo; i++)
			printf("%s %d\n", mi[i], mv[i]);
	}
	
	return 0;
}
